File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(heights) {
    if (heights.length < 3) return 0;
    
    let [maxLeft, maxRight] = [heights[0], heights[heights.length - 1]];
    let [left, right] = [1, heights.length - 2];
    
    let totalRain = 0;
    
    while (left <= right) {
        if (maxLeft < maxRight) {
            if (heights[left] > maxLeft) maxLeft = heights[left];
            else totalRain += maxLeft - heights[left];
            left++;
        } else {
            if (heights[right] > maxRight) maxRight = heights[right];
            else totalRain += maxRight - heights[right];
            right--;
        }
    }
    
    return totalRain;
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, heights: list[int]) -> int:
        total_rain = 0
        left, right = 0, len(heights) - 1

        while left < right:
            l_height = heights[left]
            r_height = heights[right]

            if l_height < r_height:
                if l_height < heights[left + 1]:
                    left += 1
                else:
                    while left < right and not l_height < heights[left + 1]:
                        left += 1
                        total_rain += l_height - heights[left]
            else:
                if r_height < heights[right - 1]:
                    right -= 1
                else:
                    while left < right and not r_height < heights[right - 1]:
                        right -= 1
                        total_rain += r_height - heights[right]

        return total_rain